tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,229 @@
+
+  const MappingOptionsType& bestMappingOptions() {
+    std::lock_guard<std::mutex> lock(bestTimeMutex_);
+    return bestMappingOptions_;
+  }
+
+ private:
+  /// Synchronization related objects
+  /// The main invariant is that we always try to compile and evaluate
+  /// exactly searchStrategy->population.size() candidates.
+  /// If a candidate fails compilation we still add a null Executor so that
+  /// the invariant holds.
+  /// This way it is easy to implement multi-threaded termination by just
+  /// taking an atomic counter and pushing/popping the queues under lock until
+  /// we have evaluated searchStrategy->population.size() compilation results.
+  std::mutex bestTimeMutex_;
+  std::mutex executorsMutex_;
+  std::atomic_bool stopRequested_;
+  std::atomic_size_t currentCompilationJob_;
+  std::atomic_size_t numEvaluations_;
+  std::queue<std::unique_ptr<ExecutorType>> executors_;
+  std::queue<CandidateConfiguration*> configurations_;
+
+  /// inputs
+  lang::TreeRef kTcTree_;
+  const MappingOptionsType kBaseMapping_;
+  /// maps of inputs and outputs per each device (represented by a size_t)
+  /// involved in autotuning. The client of the autotuner API must allocate
+  /// these properly on each device where autotuning evaluation needs to run.
+  /// In particular all the inputs and outputs must contain the same values
+  /// across devices for the purpose of running correctness checks during
+  /// tuning (future work).
+  const std::unordered_map<size_t, std::vector<const DLConstTensor*>> kInputs_;
+  std::unordered_map<size_t, std::vector<const DLTensor*>> outputs_;
+
+  // results
+  size_t bestTime_;
+  MappingOptionsType bestMappingOptions_;
+
+  // backing options cache
+  std::shared_ptr<OptionsCache<Backend>> optionsCache_;
+};
+} // namespace detail
+
+/**
+ * An Autotuner provides the basic interface to run a SearchStrategy over a
+ * particular Backend.
+ * Under the hood it instantiates a TuningHarness<Backend> and a
+ * SearchStrategy. The Autotuner then calls TuningHarness<Backend>::run()
+ * on the SearchStrategy.
+ */
+template <typename Backend, typename SearchStrategy>
+class Autotuner {
+ public:
+  using MappingOptionsType = typename Backend::MappingOptionsType;
+  using OptionsCacheType = tc::autotune::OptionsCache<Backend>;
+
+  /// Parse and store the underlying tree representation for all the TC
+  /// functions provided.
+  explicit Autotuner(const std::string& tc);
+
+  /// Runs autotuning on the TC function tcEntryPoint.
+  /// This assumes input and output tensors of proper sizes have been
+  /// initialized on each device. For now this initialization is left to the
+  /// client of the Autotuner.
+  /// Tuning performs a search of template type SearchStrategy starting from
+  /// baseMapping.
+  ///
+  /// Alternatively an OptionsCache cacheFileName serialized path
+  /// can be specified from which the tuner recovers multiple starting points
```

Comment:
This seems to only be true if a specific CLI flag is set...

Reply:
killing that flag with fire